### ISSUE TITLE
Property/Action/Emitter/Tiny* class hierachy

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/axon#343]
The relationships beweem Property, Action, Emitter, and Tiny* are really odd, to put it kindly.  

The latest casualty is https://github.com/phetsims/twixt/issues/29, where Node `opacityProperty` cannot be animated because it's not actually a Property.   So I guess that API is going to have to resort to something like:

```js
assert && assert( 
  ( options.property instanceof Property || options.property.instanceof TinyProperty ) &&
  options.property.isSettable(), 
  '...' 
);
```

The relationships are so complicated and non-intuitive that I had to make a diagram to wrap my head around the big picture.  Here's the big picture as of 3/3/2021.  This shows the class hierachies using indentation, and composition relationships as denoted by `hasA`. 

```
TinyEmitter
  Timer
  TinyProperty - isSettable is true
    TinyForwardingProperty
    TinyStaticProperty
    
PhetioObject
  Property - hasA TinyEmitter, isSettable is true
    BooleanProperty
    ColorProperty
    DerivedProperty - isSettable is false
    DynamicProperty - isSettable is conditional
    EnumerationProperty
    NodeProperty
    NumberProperty
    PaintColorProperty
    StringProperty
    Vector2Property
 
PhetioObject   
  Action
    Emitter - hasA TinyEmitter
```

Amazing to me is that TinyProperty is not at all related to Property, and Property is not backed by a TinyProperty via composition.  At least Emitter is backed by TinyEmitter.  But I digress... I'm pretty sure I could throw stones at this for hours, but that's probably not going to be productive.  

I honestly don't know what (if anything) to do about this. Unfortunately this has been grown,  not designed. And these classes are literally everywhere, so changing things would be a herculean task.  Maybe we just continue in the direction that we're going, and end up having to specify `{Property|TinyProperty}` and/or check `isSettable` everywhere.  That's a pretty bleak "solution" for something that's PhET's foundation.

**Comment by pixelzoom** [comment]
I have one concrete suggestion for improvement in https://github.com/phetsims/axon/issues/344. Property should use TinyProperty, to eliminate duplicate code that has been copied from Property to TinyProperty, and to establish a coherent composition (`hasA`) relationship between these classes.

**Comment by zepumph** [comment]
Javascript doesn't have interfaces. we all wish it did. We discussed this issue with Property last over in https://github.com/phetsims/sun/issues/641. 

We are going to improve the Action/Emitter hierarchy over in https://github.com/phetsims/phet-io/issues/1543. I would much prefer that refactor be limited unless we talk about it in the context of that issue, and the many ideas towards the bottom of it.

I would be ok with (a) creating mixins as our Property/Emitter interfaces or (b) documenting these as jsdoc interfaces. In past discussions no one has seemed too keen on either of those. We need Tiny* to be as bare bones as they are or we don't get to use them in Scenery. 

We could also create PropertyDef as a typedef/interface and use something like `PropertyDef.isProperty` for assertions and forbid `instanceof Property` checks.

I agree #344 would be nice.





>I honestly don't know what (if anything) to do about this. Unfortunately this has been grown, not designed. And these classes are literally everywhere, so changing things would be a herculean task. Maybe we just continue in the direction that we're going, and end up having to specify {Property|TinyProperty} and/or check isSettable everywhere. That's a pretty bleak "solution" for something that's PhET's foundation.

I totally agree, but also want to add that I feel like we had this developer meeting discussion only recently over in https://github.com/phetsims/phet-io/issues/1543 at the end of January. We did the best we could there with the addition of `isSettable` to unblock an issue. To me it would be most helpful if we could frame this conversation as fixes to specific problems, because we have already had a "wouldn't it be nice" conversation two months ago, and agreed that there wasn't the bandwidth to dive into that right now.

**Comment by pixelzoom** [comment]
>  I feel like we had this developer meeting discussion only recently over in phetsims/phet-io#1543 at the end of January.

No, we had a subset of this discussion.  Our discussion was focused on the fact that some subclasses of Property are not settable.  I had no idea what the picture in https://github.com/phetsims/axon/issues/343#issue-821581997 looked like at the time, and no awareness of the "TinyProperty is not a Property" problem.  We have never addressed the issue of what should happen when a TinyProperty is passed to an API that expects a settable Property.  If the API is doing type checking, it will currently fail.  This is exactly the problem that I bumped into today in phetsims/twixt#29.   

> We did the best we could there with the addition of `isSettable` to unblock an issue. 

`isSettable` was added on 1/26/21.  It is currently being used zero times in the entire PhET code base. Nice that we were able to unblock the issue, but it has resulted in zero impact on the problem.  

> To me it would be most helpful if we could frame this conversation as fixes to specific problems

Agreed.  I've tried to do that with #344. I'll continue to think about other ways to make this all more coherent.

**Comment by samreid** [comment]
In today's discussion, we considered:

* A typedef
* Pretending (or documenting) that TinyProperty implements the Property interface so that we would still annotate types as Property
* Investigating typescript

We do not have a consensus yet.

**Comment by pixelzoom** [comment]
More notes from 4/15/21 dev  meeting.

* There was a bit of cleanup done to the hierarchies in #344.  `Property` now hasA `TinyProperty`.  This is a nice improvement, and provides symmetry with `Emitter` and `TinyEmitter`.

* If this were something other than Property et.al., we'd probably close this issue and move on.  But Property et.al. are used widely in PhET-iO APIs.

To summarize the problem...  Because of the odd class hierarchy and hasA relationships,  we are not able to specify 2 important things for an API:

(1) The correct type expression for a`@param` or `option`
(2) Whether the `@param` or `option` is settable

Most PhET APIs look something like this:

```js
/**
 * @param {Property.<...>} someProperty
 */
someFunction( someProperty ) 
```

General problems:
* The type expression does not include the `TinyProperty` hierarchy, but may be intended.
* The type expression does indicate whether `someProperty` must be settable.
* Tthere is typically no `someProperty.isSettable()` assertion when there should be. (There are 2 uses of `isSettable` in the entire PhET code base.)
* Type checking of `someProperty` can be problematic and complicated.

**Comment by samreid** [comment]
I recommend discussing and scoping out the possibility of a Typescript migration before spending much more time on this issue.

**Comment by jessegreenberg** [comment]
Developer meeting 4/29/21 - 

It seems like the solution to this would involved Typescript, that is being investigated in https://github.com/phetsims/tasks/issues/987. This is on hold until we know how we will proceed in that issue.

**Comment by samreid** [comment]
As mentioned in https://github.com/phetsims/axon/issues/363, the base class or interface should be read-only, then the subclasses should add on the ability to reset, set value, etc.

**Comment by jonathanolson** [comment]
With typescript, can't we create separate interfaces for read-only, write-only, and read-write? Do we necessarily have to solve this with hierarchy, or can we solve this with the type system?

**Comment by zepumph** [comment]
Is this a duplicate of https://github.com/phetsims/axon/issues/404?

**Comment by samreid** [comment]
Yes, that issue is solving the same problem and shows much more progress.  Closing, but anyone including @pixelzoom feel free to reopen if there are remarks above that still need attention.
